pkg/common/patcher/patcher.go

```diff
@@ -0,0 +1,157 @@
+package patcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	jsonpatch "github.com/evanphx/json-patch"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Patcher is just the Patch API with a generic to keep use sites type safe.
+// This is inspired by the commiter code in https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/committer/committer.go
+type PatchClient[R runtime.Object] interface {
+	Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (R, error)
+}
+
+type Patcher[R runtime.Object, Sp any, St any] interface {
+	AddFinalizer(context.Context, R, string) (bool, error)
+	RemoveFinalizer(context.Context, R, string) error
+	PatchStatus(context.Context, R, St, St) (bool, error)
+	PatchSpec(context.Context, R, Sp, Sp) (bool, error)
+}
+
+// Resource is a generic wrapper around resources so we can generate patches.
+type Resource[Sp any, St any] struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              Sp `json:"spec"`
+	Status            St `json:"status,omitempty"`
+}
+
+type patchFunc[R runtime.Object, Sp any, St any] func(context.Context, R, *Resource[Sp, St], *Resource[Sp, St], ...string) error
```

Comment:
I do not see this being used in other places except line 39.

Reply:
oops we do not